src/main/scala/org/ergoplatform/nodeView/wallet/WalletScanLogic.scala

```diff
@@ -63,46 +66,91 @@ object WalletScanLogic extends ScorexLogging {
                             walletVars: WalletVars,
                             height: Int,
                             blockId: ModifierId,
-                            transactions: Seq[ErgoTransaction]): (WalletRegistry, OffChainRegistry) = {
-
-    //todo: inefficient for wallets with many outputs, replace with a Bloom/Cuckoo filter?
-    //todo: at least, cache in RAM instead of reading from DB
-    val previousBoxIds = registry.allUnspentBoxes().map(tb => encodedBoxId(tb.box.id))
+                            transactions: Seq[ErgoTransaction],
+                            cachedOutputsFilter: Option[BloomFilter[Array[Byte]]]
+                           ): (WalletRegistry, OffChainRegistry, BloomFilter[Array[Byte]]) = {
+
+    // Take unspent wallet outputs Bloom Filter from cache
+    // or recreate it from unspent outputs stored in the database
+    val outputsFilter = cachedOutputsFilter.getOrElse {
+      // todo: currently here and other places hardcoded values are used for Bloom filter parameters
+      // todo: consider different profiles for the config, such as "user", "wallet", "app hub"
+      val expectedKeys = 20000
+      val falsePositiveRate = 0.01
+      val bf = BloomFilter.create[Array[Byte]](Funnels.byteArrayFunnel(), expectedKeys, falsePositiveRate)
+
+      registry.allUnspentBoxes().foreach { tb =>
+        bf.put(tb.box.id)
+      }
+      bf
+    }
 
+    // Resolve boxes related to mining income
     val resolvedBoxes = registry.unspentBoxes(MiningScanId).flatMap { tb =>
       val spendable = resolve(tb.box, walletVars.proverOpt, stateContext, height)
       if (spendable) Some(tb.copy(scans = Set(PaymentsScanId))) else None
     }
 
     //input tx id, input box id, tracked box
-    type InputData = Seq[(ModifierId, EncodedBoxId, TrackedBox)]
+    type InputData = Seq[(ModifierId, ModifierId, TrackedBox)]
```

Comment:
Consider introducting case class instead of the tuple

Reply:
done